### ISSUE TITLE
KeyError when trying to create new postgres user with salt state

### ISSUE THREAD

**Issue by midihenry** [saltstack/salt#27321]
When trying to create a new postgresql user using the following state
```saltstack
...
add_user:
  postgres_user.present:
    - name: {{ postgres.db_user }}
    - password: {{ postgres.db_password }}
    - user: postgres
    - require:
      - pkg: postgresql_install
...
```
I get this error
```python-traceback
The minion function caused an exception: Traceback (most recent call last):
      File "/usr/lib/python2.7/dist-packages/salt/minion.py", line 1161, in _thread_return
        return_data = func(*args, **kwargs)
      File "/usr/lib/python2.7/dist-packages/salt/modules/state.py", line 523, in highstate
        whitelist=kwargs.get('whitelist')
      File "/usr/lib/python2.7/dist-packages/salt/state.py", line 2968, in call_highstate
        return self.state.call_high(high)
      File "/usr/lib/python2.7/dist-packages/salt/state.py", line 2067, in call_high
        ret = dict(list(disabled.items()) + list(self.call_chunks(chunks).items()))
      File "/usr/lib/python2.7/dist-packages/salt/state.py", line 1623, in call_chunks
        running = self.call_chunk(low, running, chunks)
      File "/usr/lib/python2.7/dist-packages/salt/state.py", line 1769, in call_chunk
        self._mod_init(low)
      File "/usr/lib/python2.7/dist-packages/salt/state.py", line 612, in _mod_init
        self.states['{0}.{1}'.format(low['state'], low['fun'])]  # pylint: disable=W0106
      File "/usr/lib/python2.7/dist-packages/salt/utils/lazy.py", line 90, in __getitem__
        raise KeyError(key)
    KeyError: 'postgres_user.present'

```

I'm using salt-master version 2015.5.5 and 2015.5.3 for the salt-minion

**Comment by jfindlay** [comment]
@midihenry, thanks for the report.  What is the output from `salt minion sys.doc postgres.user_exists` and `salt minion cmd.run 'which psql'`?

**Comment by kitplummer** [comment]
Seeing the same on CentOS6.  The error happens before I can even get Postgres installed, nothing actually runs with a salt-call.

**Comment by jfindlay** [comment]
@kitplummer, when you say that it happens before you can get it installed, do you mean that you install and then add a postgres user in the same state run?  Is the user state guaranteed to happen after the install state by using requisites as necessary?  Is it possible for you to post your state file(s)?  Thanks.

**Comment by kitplummer** [comment]
Here's my state.  It was working.

https://gist.github.com/kitplummer/d3f4f821f3dba7a47aa8

**Comment by jfindlay** [comment]
This seems to be an issue with the `postgresql-install` state in that the loader is not reloading modules between function calls.  I think this format is supported, but I can only surmise this from reading states from users.  @basepi, will probably know much better than I.  If this is a bug, it is likely similar to if not the same as #27016 and another issue that I've seen but can't find.

**Comment by basepi** [comment]
> It was working.

What changed? Different version of salt? Small changes to the state itself?

**Comment by kitplummer** [comment]
Just upgrades of Salt, master and minion.

**Comment by basepi** [comment]
Can you add `- pkg: postgresql-install` to the requires for the state that's failing? I want to make sure it happens after the packages are installed.

**Comment by kitplummer** [comment]
Sure, I tried just that.

```
# salt minion state.apply postgresql
minion:
    The minion function caused an exception: Traceback (most recent call last):
      File "/usr/lib/python2.6/site-packages/salt/minion.py", line 1173, in _thread_return
        return_data = func(*args, **kwargs)
      File "/usr/lib/python2.6/site-packages/salt/modules/state.py", line 308, in apply_
        return sls(mods, **kwargs)
      File "/usr/lib/python2.6/site-packages/salt/modules/state.py", line 708, in sls
        ret = st_.state.call_high(high_)
      File "/usr/lib/python2.6/site-packages/salt/state.py", line 2068, in call_high
        ret = dict(list(disabled.items()) + list(self.call_chunks(chunks).items()))
      File "/usr/lib/python2.6/site-packages/salt/state.py", line 1624, in call_chunks
        running = self.call_chunk(low, running, chunks)
      File "/usr/lib/python2.6/site-packages/salt/state.py", line 1770, in call_chunk
        self._mod_init(low)
      File "/usr/lib/python2.6/site-packages/salt/state.py", line 613, in _mod_init
        self.states['{0}.{1}'.format(low['state'], low['fun'])]  # pylint: disable=W0106
      File "/usr/lib/python2.6/site-packages/salt/utils/lazy.py", line 90, in __getitem__
        raise KeyError(key)
    KeyError: 'postgres_user.present'
```

**Comment by basepi** [comment]
Does `salt 'myminion' postgres.user_exists <user>` work for you? Or does it report as "not available"?

**Comment by kitplummer** [comment]
`Module 'postgres' is not available.
ERROR: Minions returned with non-zero exit code`

**Comment by basepi** [comment]
Alright, so there are two things going on. One is that the `postgres` module is not loading properly. The only requirement for loading this module is that the module can find `psql` on the command line. Is that not being installed, or being installed to a nonstandard location?

It is weird, however, that that KeyError is getting through in the state system. It should just report that `postgres_user.present` is unavailable, much like it does with the module. May be a separate bug there.

**Comment by midihenry** [comment]
I'm sorry for the delay but I had to move on to work on something else.

I can confirm that the state file was working before but now it doesn't anymore. 
After running the following state file:
```
postgresql_install:
  pkg.installed:
    - name: postgresql-9.3
    - refresh: True

add_user:
  postgres_user.present:
    - name: user
    - password: user
    - user: postgres
    - require:
      - pkg: postgresql_install

```
this is what I get when opening the salt-minion log file:
![minion_log](https://cloud.githubusercontent.com/assets/7854083/10165751/c1286bb8-66b8-11e5-84ca-43c40b062550.png)

So I thought it could be a missing repo maybe? 
(this happened on ubuntu 14.04 server edition)

**Comment by basepi** [comment]
It does appear to be an issue with a missing dependency, rather than a bug in salt. I guess your next step is to figure out how to get `postgresql-common` installed, and then we can decide if salt should have taken steps that it didn't.

**Comment by kitplummer** [comment]
For me, 'psql' is def missing.  It looks like I have an upstream Yum repo
issue.  Perhaps Salt could just be lil nicer with some kind of error
handling.

**Comment by basepi** [comment]
Yes, the fact that it's showing a stacktrace is odd, it should just be reporting the state module as "not available"...

**Comment by midihenry** [comment]
It was a repo problem indeed (for me on ubuntu 14.04). 
Postgresql couldn't be installed even with a simple apt-get install. I was finally able to get the salt state running without a problem after adding the following repos to /etc/apt/sources.list:
```
deb http://ftp.de.debian.org/debian/ wheezy main non-free contrib
deb-src http://ftp.de.debian.org/debian/ wheezy main non-free contrib
```

**Comment by basepi** [comment]
Awesome! I'm glad we tracked it down.

**Comment by N-Mi** [comment]
I have the same issue on a fresh new Debian Jessie minion.

After doing a "aptitude install postgres-client" directly on the minion, things went way better, and state applied correctly.

**Comment by midihenry** [comment]
But shouldn't the state return an error before moving to the postgres user creation? 
If postgresql installation fails, and it's a requirement for running the user creation, shouldn't we get an error of the type "one or more requisites failed" ?

**Comment by basepi** [comment]
Was the install state actually showing failure? I think it was not, which leads me to believe there may be a bug in the aptpkg.py pkg handling.

**Comment by midihenry** [comment]
@basepi No the install state was not showing failure. It fails silently.
The only error I got was the KeyError exception:
```
The minion function caused an exception: Traceback (most recent call last):
      File "/usr/lib/python2.7/dist-packages/salt/minion.py", line 1161, in _thread_return
        return_data = func(*args, **kwargs)
      File "/usr/lib/python2.7/dist-packages/salt/modules/state.py", line 523, in highstate
        whitelist=kwargs.get('whitelist')
      File "/usr/lib/python2.7/dist-packages/salt/state.py", line 2968, in call_highstate
        return self.state.call_high(high)
      File "/usr/lib/python2.7/dist-packages/salt/state.py", line 2067, in call_high
        ret = dict(list(disabled.items()) + list(self.call_chunks(chunks).items()))
      File "/usr/lib/python2.7/dist-packages/salt/state.py", line 1623, in call_chunks
        running = self.call_chunk(low, running, chunks)
      File "/usr/lib/python2.7/dist-packages/salt/state.py", line 1769, in call_chunk
        self._mod_init(low)
      File "/usr/lib/python2.7/dist-packages/salt/state.py", line 612, in _mod_init
        self.states['{0}.{1}'.format(low['state'], low['fun'])]  # pylint: disable=W0106
      File "/usr/lib/python2.7/dist-packages/salt/utils/lazy.py", line 90, in __getitem__
        raise KeyError(key)
    KeyError: 'postgres_user.present'
```

**Comment by basepi** [comment]
@midihenry Could you please file a new issue just for the fact that pkg.installed wasn't failing even though apt was reporting missing dependencies? We should definitely resolve that.

**Comment by midihenry** [comment]
Will do :+1:

**Comment by rallytime** [comment]
Hi everyone! These stacktraces that end in KeyErrors like this this with the following:
```
File "/usr/lib/python2.7/dist-packages/salt/state.py", line 1769, in call_chunk
        self._mod_init(low)
      File "/usr/lib/python2.7/dist-packages/salt/state.py", line 612, in _mod_init
        self.states['{0}.{1}'.format(low['state'], low['fun'])]  # pylint: disable=W0106
      File "/usr/lib/python2.7/dist-packages/salt/utils/lazy.py", line 90, in __getitem__
        raise KeyError(key)
    KeyError: 'postgres_user.present'
```
Are a known issue that had been fixed in 2015.5.6 and 2015.8.0. We had a regression in the state compiler causing this same stacktrace in many bug reports filed against 2015.5.3. 

The PRs responsible for fixing this are #25928 for the 2015.8 branch, which was back-ported to the 2015.5 branch with #27681.

Can anyone test these PRs, releases/tags, or the HEAD of these branches and confirm that this is fixed?

**Comment by rallytime** [comment]
Oh also, @midihenry the KeyError you're hitting in your most recent comment is also likely this same error. Did you ever open a new issue? Because we should track that one as being fixed if so.

**Comment by midihenry** [comment]
@rallytime Yes I did.
This is the issue that I opened regarding the error: https://github.com/saltstack/salt/issues/27619
It seems that you've already checked it out :)

**Comment by rallytime** [comment]
Oh, ha! Thanks @midihenry. I look at so many issues, I didn't realize I'd already seen that one. Have you had a chance to test this solution? Do you think we can close either of these issues?

**Comment by midihenry** [comment]
@rallytime Unfortunately, I didn't have the time to test it again as I've been working on something else the whole time. 
However, I'll try to test it as soon as I can.

**Comment by rallytime** [comment]
No problem! Let us know what you find out when you get a chance. ;)

**Comment by rallytime** [comment]
Update: It appears that I mis-remembered when my back-porting PR went into the 2015.5 branch and it missed the 2015.5.6 release tag. However, the initial stacktrace reported by @midihenry about the KeyError should be fixed at the HEAD of the 2015.5 branch. The fix being in 2015.8.0 is still correct, however.

**Comment by midihenry** [comment]
Thanks for the update @rallytime !

**Comment by rallytime** [comment]
Since I have confirmed that this is fixed on the 2015.5 branch and the fix will be available in the 2015.5.7 release, I am going to close this. If it comes up again for anyone, please leave a comment and we can reopen this issue. Thanks!
